packages/@aws-cdk/aws-codedeploy/lambda-packages/ecs-deployment-provider/lib/on-event.ts

```diff
@@ -0,0 +1,116 @@
+
+const logger = new Logger({ serviceName: 'ecsDeploymentProviderOnEvent' });
+const codedeployClient = new CodeDeploy({});
+
+/**
+ * The properties of the CodeDeploy Deployment to create
+ */
+interface DeploymentProperties {
+  description: string;
+  applicationName: string;
+  deploymentConfigName: string;
+  deploymentGroupName: string;
+  autoRollbackConfigurationEnabled: string;
+  autoRollbackConfigurationEvents: string;
+  revisionAppSpecContent: string;
+}
+
+/**
+ * The properties in the Data object returned to CloudFormation 
+ */
+export interface DataAttributes {
+  deploymentId: string;
+}
+
+/**
+ * The request object that the custom resource lamba function receives from CloudFormation.
+ */
+export interface OnEventRequest {
+  RequestType: string;
+  LogicalResourceId: string;
+  PhysicalResourceId: string;
+  ResourceProperties: DeploymentProperties;
+  OldResourceProperties: DeploymentProperties;
+  ResourceType: string;
+  RequestId: string;
+  StackId: string;
+}
+/**
+ * The response object that the custom resource lambda function returns to CloudFormation.
+ */
+export interface OnEventResponse {
+  PhysicalResourceId?: string;
+  Data?: DataAttributes;
+  NoEcho?: boolean;
+}
+
+/**
+ * The lambda function called from CloudFormation for this custom resource.
+ * 
+ * @param event 
+ * @returns attribues of the deployment that was created
+ */
+export async function handler(event: OnEventRequest): Promise<OnEventResponse> {
+  logger.appendKeys({
+    stackEvent: event.RequestType,
+  });
+  switch (event.RequestType) {
+    case 'Create':
+    case 'Update': {
+      // create deployment
+      const props = event.ResourceProperties;
+      const resp = await codedeployClient.createDeployment({
+        applicationName: props.applicationName,
+        deploymentConfigName: props.deploymentConfigName,
+        deploymentGroupName: props.deploymentGroupName,
+        autoRollbackConfiguration: {
+          enabled: props.autoRollbackConfigurationEnabled === 'true',
+          events: props.autoRollbackConfigurationEvents?.split(','),
+        },
+        description: props.description,
+        revision: {
+          revisionType: 'AppSpecContent',
+          appSpecContent: {
+            content: props.revisionAppSpecContent,
+          },
+        },
+      });
+      if (!resp.deploymentId) {
+        throw new Error('No deploymentId received from call to CreateDeployment');
+      }
+      logger.appendKeys({
+        deploymentId: resp.deploymentId,
+      });
+      logger.info('Created new deployment');
+
+      return {
+        PhysicalResourceId: resp.deploymentId,
+        Data: {
+          deploymentId: resp.deploymentId,
+        },
+      };
+    }
+    case 'Delete':
+      logger.appendKeys({
+        deploymentId: event.PhysicalResourceId,
+      });
+      // cancel deployment and rollback
+      try {
+        const resp = await codedeployClient.stopDeployment({
```

Comment:
Should we first look to see if the deployment with the given `deploymentId` is in process? If it isn't then there is nothing to do.

If there is an in progress deployment that we stop, I think we should then track the progress of that.

Reply:
> Should we first look to see if the deployment with the given deploymentId is in process? If it isn't then there is nothing to do.

This can be unreliable in the event of race conditions. Better to try to stop and handle situations where there isn't a deployment in progress as a success and final event.

> If there is an in progress deployment that we stop, I think we should then track the progress of that.

This is already properly handled in the `is-complete.ts` handler.